language/diem-vm/mvhashmap/src/lib.rs

```diff
@@ -0,0 +1,235 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use once_cell::sync::OnceCell;
+use std::{
+    cmp::{max, PartialOrd},
+    collections::{btree_map::BTreeMap, HashMap},
+    fmt::Debug,
+    hash::Hash,
+};
+
+#[cfg(test)]
+mod unit_tests;
+
+/// A structure that holds placeholders for each write to the database
+//
+//  The structure is created by one thread creating the scheduling, and
+//  at that point it is used as a &mut by that single thread.
+//
+//  Then it is passed to all threads executing as a shared reference. At
+//  this point only a single thread must write to any entry, and others
+//  can read from it. Only entries are mutated using interior mutability,
+//  but no entries can be added or deleted.
+//
+
+pub type Version = usize;
+
+pub struct MVHashMap<K, V> {
+    data: HashMap<K, BTreeMap<Version, WriteCell<V>>>,
+}
+
+#[derive(Debug)]
+pub enum Error {
+    UnexpectedWrite,
+}
+
+#[cfg_attr(any(target_arch = "x86_64"), repr(align(128)))]
+pub(crate) struct WriteCell<V>(OnceCell<Option<V>>);
+
+impl<V: Debug> WriteCell<V> {
+    pub fn new() -> WriteCell<V> {
+        WriteCell(OnceCell::new())
+    }
+
+    pub fn is_assigned(&self) -> bool {
+        self.0.get().is_some()
+    }
+
+    pub fn write(&self, v: V) {
+        self.0
+            .set(Some(v))
+            .expect("WriteCell should be modified exactly once");
+    }
+
+    pub fn skip(&self) {
+        self.0
+            .set(None)
+            .expect("WriteCell should be modified exactly once");
+    }
+
+    pub fn get(&self) -> Option<&Option<V>> {
+        self.0.get()
+    }
+}
+
+impl<K: Hash + Clone + Eq, V: Clone + Debug> MVHashMap<K, V> {
+    /// Create the MVHashMap structure from a list of possible writes. Each element in the list
+    /// indicates a key that could potentially be modified at its given version.
+    ///
+    /// Returns the maximum number of writes that can write to one single key, and the MVHashMap.
+    pub fn new_from(possible_writes: Vec<(K, Version)>) -> (usize, MVHashMap<K, V>) {
```

Comment:
super nit: doesn't it make sense for the first return value to be HashMap since that's what we are constructing?